Util/src/main/java/io/deephaven/util/reference/CleanupReferenceProcessor.java

```diff
@@ -132,33 +208,138 @@ public final synchronized void resetForUnitTests() {
+            return (ReferenceQueue<T>) referenceQueue;
+        }
+
+        public <T> CleanupReference<T> register(T obj, Runnable action) {
+            final PhantomCleanupRef<T> ref = new PhantomCleanupRef<>(obj, referenceQueue(), action);
+            registrations.add(ref);
+            return ref;
+        }
 
-        private DrainQueue(ReferenceQueue<?> localQueue) {
-            this.localQueue = localQueue;
+        public <T> CleanupReference<T> registerWeak(T obj, Runnable action) {
+            final WeakCleanupRef<T> ref = new WeakCleanupRef<>(obj, referenceQueue(), action);
+            registrations.add(ref);
+            return ref;
+        }
+
+        public <T> CleanupReference<T> registerSoft(T obj, Runnable action) {
+            final SoftCleanupRef<T> ref = new SoftCleanupRef<>(obj, referenceQueue(), action);
+            registrations.add(ref);
+            return ref;
         }
 
         @Override
         public void run() {
-            while (localQueue == referenceQueue) {
+            while (this == drainQueue) {
                 final Reference<?> reference;
                 try {
-                    reference = localQueue.remove(shutdownCheckDelayMillis);
+                    reference = referenceQueue.remove(shutdownCheckDelayMillis);
                 } catch (InterruptedException ignored) {
                     continue;
                 }
                 if (reference instanceof CleanupReference) {
+                    final CleanupReference<?> ref = (CleanupReference<?>) reference;
                     try {
                         if (LOG_CLEANED_REFERENCES) {
                             log.info().append("CleanupReferenceProcessor-").append(name).append(", cleaning ")
                                     .append(Utils.REFERENT_FORMATTER, reference).endl();
                         }
-                        ((CleanupReference<?>) reference).cleanup();
+                        ref.cleanup();
                     } catch (Exception e) {
-                        exceptionHandler.accept(log, (CleanupReference<?>) reference, e);
+                        exceptionHandler.accept(log, ref, e);
+                    }
+                    if (ref instanceof RegisteredCleanupReference) {
+                        registrations.remove(ref);
                     }
                 }
             }
         }
     }
+
+    interface RegisteredCleanupReference<T> extends CleanupReference<T> {
+
+    }
+
+    private static class PhantomCleanupRef<T> extends PhantomReference<T> implements RegisteredCleanupReference<T> {
+        private Runnable action;
+
+        PhantomCleanupRef(T referent, ReferenceQueue<? super T> q, Runnable action) {
+            super(referent, q);
+            this.action = Objects.requireNonNull(action);
+            Reference.reachabilityFence(referent);
+            Reference.reachabilityFence(q);
+        }
+
+        @Override
+        public void cleanup() {
```

Comment:
You *could* use interface defaulting to avoid this duplication....

Reply:
If there was more complex logic, I would consider it. In this case, I think per-impl is pretty straightforward.